### DISCUSSION TITLE
(NetApp StorageGRID) S3-Check fails since v0.22.0

### DISCUSSION THREAD

**Discussion by ganonym** [pocketbase/pocketbase#4806] [Q&A] [answered]
Hello everyone,

I currently have the problem that when updating from 0.21.* to 0.22.* the integrated S3 storage can no longer be accessed.

When trying to establish a connection, the following error message always appears:

![image](https://github.com/pocketbase/pocketbase/assets/159258586/2859222d-43bf-437a-9446-57f049a87a0f)

The error appears from version 0.22.0, where there is also a note in the release that the connection to the S3 storage should be checked again.

Is this an error on my part?
Or has something been removed/changed in version 0.22.0 that is causing this error?

**Comment by ganigeorgiev** [comment]
What S3 provider do you use?

The above error indicates that either the S3 provider is incompatible or there is an issue with the S3 permissions. In both cases it doesn't explain why it would work with v0.21 and not v0.22.
There was a small change to the S3 driver in [v0.22.7](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.7) but the above error doesn't seem related to it and I've tested it with Wasabi and the MinIO playground and it is working correctly.

**Reply by ganigeorgiev** [reply]
I guess it could be maybe related to the upgrade to `aws-sdk-go-v2` with [v0.22.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.0) but without some information about the used provider I'm not sure how to help.

**Reply by ganonym** [reply]
I use the S3 storage from a NetApp StorageGRID. The S3 storage is based on ONTAP S3.
I can actually rule out an incompatibility, as a connection to the S3 storage (with PB versions below 0.22.0) could always be established.

I have tested all other versions from 0.22.0 to 0.22.9 and the same error occurs with all of them.

**Reply by ganigeorgiev** [reply]
It is most likely caused by the upgrade to `aws-sdk-go-v2` but I'm not sure how to debug it as I'm not familiar with the NetApp offerings.

From quick search of the error there seems to be several articles with similar error message to yours:
- https://kb.netapp.com/onprem/ontap/da/S3/Multipart_S3_upload_not_implemented
- https://kb.netapp.com/onprem/ontap/os/Javascript_AWS_SDK_v3_requests_fails_with%3A_not_implemented
- https://kb.netapp.com/onprem/ontap/da/S3/MinIO_client_list_ONTAP_S3_bucket_fails_with_error_header_you_provided_implies_functionality_that_is_not_implemented

_(I'm not able to read them because they are behind an auth wall)_

I'm not sure what has changed between `aws-sdk-go` and `aws-sdk-go-v2` that will cause this error (for GCS there was difference in the signing headers but the result error was 40x related, not 501).

NetApp doesn't seem to have a free trial (or at least one that doesn't require contacting their sales team) so I'm not sure how to investigate it (_I've tried to register but it seems that I have to wait to be approved_).

You can try creating a minimal reproducible example with just the [`aws-sdk-go-v2`](https://github.com/aws/aws-sdk-go-v2) and try to ask for help in the related repo.

**Reply by ganigeorgiev** [reply, marked as answer]
I haven't tested it but a minimal `aws-sdk-go-v2` example could look like:

```go
// main.go

package main

import (
	"bytes"
	"context"
	"log"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// replace with your settings
const (
	accessKey  = "..."
	secretKey  = "..."
	region     = "..."
	endpoint   = "..."
	bucketName = "..."
)

func main() {
	ctx := context.Background()

	cred := credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")

	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithCredentialsProvider(cred),
		config.WithRegion(region),
		config.WithEndpointResolverWithOptions(aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
			return aws.Endpoint{URL: endpoint, SigningRegion: region}, nil
		})),
	)
	if err != nil {
		log.Fatal(err)
	}

	s3Client := s3.NewFromConfig(cfg)

	_, err = s3Client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String("test_file"),
		Body:   bytes.NewReader([]byte("test123")),
	})
	if err != nil {
		log.Fatalf("Failed to upload file: %s", err.Error())
	}

	log.Println("Success!")
}
```
_(edit: replaced file open with just a bytes reader for simplicity)_
